components/engine_rocks/src/db.rs

```diff
@@ -0,0 +1,238 @@
+use std::ops::Deref;
+use std::path::Path;
+use std::sync::Arc;
+
+use engine_rocksdb::{
+    set_external_sst_file_global_seq_no, DBIterator, IngestExternalFileOptions, Writable, DB,
+};
+use engine_traits::{
+    DeleteRangeType, Error, IterOptions, Iterable, KvEngine, Mutable, Peekable, ReadOptions,
+    Result, WriteOptions,
+};
+use tikv_util::file::calc_crc32;
+
+use crate::options::{RocksReadOptions, RocksWriteOptions};
+use crate::util::{delete_all_in_range_cf, get_cf_handle};
+use crate::{Iterator, Snapshot, WriteBatch};
+
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct Rocks(Arc<DB>);
+
+impl Rocks {
+    pub fn from_db(db: Arc<DB>) -> Self {
+        Rocks(db)
+    }
+
+    pub fn get_sync_db(&self) -> Arc<DB> {
+        self.0.clone()
+    }
+
+    pub fn existed(path: &str) -> bool {
+        let path = Path::new(path);
+        if !path.exists() || !path.is_dir() {
+            return false;
+        }
+
+        // If path is not an empty directory, we say db exists. If path is not an empty directory
+        // but db has not been created, `DB::list_column_families` fails and we can clean up
+        // the directory by this indication.
+        fs::read_dir(&path).unwrap().next().is_some()
+    }
+}
+
+// TODO: Remove these cast methods after the engine traits are completed.
+impl AsRef<DB> for Rocks {
+    fn as_ref(&self) -> &DB {
+        self.0.deref()
+    }
+}
+
+impl AsMut<DB> for Rocks {
+    fn as_mut(&mut self) -> &mut DB {
+        Arc::get_mut(&mut self.0).unwrap()
+    }
+}
+
+impl AsRef<Rocks> for Arc<DB> {
+    fn as_ref(&self) -> &Rocks {
+        unsafe { &*(self as *const Arc<DB> as *const Rocks) }
+    }
+}
+
+impl AsMut<Rocks> for Arc<DB> {
+    fn as_mut(&mut self) -> &mut Rocks {
+        unsafe { &mut *(self as *mut Arc<DB> as *mut Rocks) }
+    }
+}
+
+impl KvEngine for Rocks {
+    type Snap = Snapshot;
+    type Batch = WriteBatch;
+
+    fn write_opt(&self, opts: &WriteOptions, wb: &WriteBatch) -> Result<()> {
+        if wb.get_db().path() != self.0.path() {
+            return Err(Error::Engine("mismatched db path".to_owned()));
+        }
+        let opt: RocksWriteOptions = opts.into();
+        self.0
+            .write_opt(wb.as_ref(), &opt.into_raw())
+            .map_err(Error::Engine)
+    }
+
+    fn write_batch(&self, cap: usize) -> WriteBatch {
+        WriteBatch::with_capacity(self.0.clone(), cap)
+    }
+
+    fn snapshot(&self) -> Snapshot {
+        Snapshot::new(self.0.clone())
+    }
+
+    fn sync(&self) -> Result<()> {
+        self.0.sync_wal().map_err(Error::Engine)
+    }
+
+    fn cf_names(&self) -> Vec<&str> {
+        self.0.cf_names()
+    }
+
+    fn delete_all_in_range_cf(
+        &self,
+        t: DeleteRangeType,
+        cf: &str,
+        start_key: &[u8],
+        end_key: &[u8],
+    ) -> Result<()> {
+        if start_key >= end_key {
+            return Ok(());
+        }
+
+        match t {
+            DeleteRangeType::WriteBatch => {
```

Comment:
I don't think `DeleteRangeType` is necessary.  For now, `delete_range` is never used. And when cleaning up a range, the caller should do like this which is really weird.
```
delete_all_in_range_cf(DeleteRangeType::DeleteFiles, cf, start, end);
delete_all_in_range_cf(DeleteRangeType::WriteBatch, cf, start, end);
```
For this implementation, it can be written like this 
```
 let handle = get_cf_handle(&self.0, cf)?;
 self.0.delete_files_in_range_cf(handle, start_key, end_key, false)?;
 delete_all_in_range_cf(&self.0, cf, start_key, end_key, false, /* use_delete_range*/);
```

Reply:
I saw a `delete_range` option, could you sure it will be never used?

Reply:
yes, `delete_range` has some potential critical bugs, so we set all to `false`. Once the feature is usable, we can simply implement it like this
```
let handle = get_cf_handle(&self.0, cf)?;
delete_all_in_range_cf(&self.0, cf, start_key, end_key, true /* use_delete_range*/)
```